From c787534e1ef1acf3f20963772b04132ec7f0ba3c Mon Sep 17 00:00:00 2001
From: Leonard Crestez <leonard.crestez@nxp.com>
Date: Mon, 14 May 2018 11:24:30 +0800
Subject: [PATCH 3786/5242] MLK-18297 drm/bridge: it6263: Sleep between status
 reads in connector detect

commit  248bfb599ffcb5a1b0753aeaf42e86f0bb73c126 from
https://source.codeaurora.org/external/imx/linux-imx.git

There are some issues with the HDMI connection on it6263 being slow to
come up. There is a hack there which reads the register multiple times
but it doesn't seem to be sufficient. This is particularly a problem
when inside a xen guest.

Improve this workaround by reading the status multiple times every few
5-10ms and returning connector_status_connected if any of those reads
sees HPDETECT.

Signed-off-by: Leonard Crestez <leonard.crestez@nxp.com>
Signed-off-by: Peng Fan <peng.fan@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/bridge/it6263.c |   12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/bridge/it6263.c b/drivers/gpu/drm/bridge/it6263.c
index 82dd389..f309380 100644
--- a/drivers/gpu/drm/bridge/it6263.c
+++ b/drivers/gpu/drm/bridge/it6263.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017 NXP
+ * Copyright 2017-2018 NXP
  *
  * This program is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License as published by the
@@ -391,12 +391,16 @@ static void it6263_reset(struct it6263 *it6263)
 	 * FIXME: We read status tens of times to workaround
 	 * cable detection failure issue at boot time on some
 	 * platforms.
+	 * Spin on this for up to one second.
 	 */
-	for (i = 0; i < 90; i++)
+	for (i = 0; i < 100; i++) {
 		regmap_read(it6263->hdmi_regmap, HDMI_REG_SYS_STATUS, &status);
+		if (status & HPDETECT)
+			return connector_status_connected;
+		usleep_range(5000, 10000);
+	}
 
-	return (status & HPDETECT) ? connector_status_connected :
-					connector_status_disconnected;
+	return connector_status_disconnected;
 }
 
 static const struct drm_connector_funcs it6263_connector_funcs = {
-- 
1.7.9.5

